chrome/content/zotero/xpcom/reader.js

```diff
@@ -962,6 +963,52 @@ class ReaderInstance {
 			return null;
 		}
 	}
+
+	/**
+	 * @typedef PDFPosition
+	 * @type {object}
+	 * @property {number} pageIndex
+	 * @property {number[][]} rects
+	*/
+	/** @typedef {import("../../../../reader/src/dom/common/lib/selector").Selector} OtherPosition*/
+	/** @typedef {PDFPosition | OtherPosition} ReaderPosition */
+	/**
+	 * @typedef ReaderSelection
+	 * @type {object}
+	 * @property {string} [sortIndex] - Sort index of the selection
+	 * @property {string} [pageLabel] - Page label of the selection. Only for PDF
+	 * @property {ReaderPosition} [position] - Position of the selection.
+	 * @property {string} [text] - Text of the selection
+	 */
+	/**
+	 * Return the last selection data
+	 * @returns {ReaderSelection} selection data
+	 * @example
+	 * Get the last selection text
+	 * ```js
+	 * let reader = Zotero.Reader._readers[0];
+	 * let selectedText = reader.getLastSelection().text || "";
+	 * ```
+	 */
+	getLastSelection() {
+		let selection = {};
+		if (this._internalReader._type === "pdf") {
+			let selectionRanges = this._internalReader._lastView._selectionRanges;
+			if (selectionRanges.length !== 0) {
+				selection ??= this._internalReader._lastView._getAnnotationFromSelectionRanges(
+					selectionRanges,
+					"highlight"
+				);
+			}
+		}
+		else {
+			selection ??= this._internalReader._lastView._getAnnotationFromTextSelection(
+				"highlight"
+			);
+		}
+		let { sortIndex, pageLabel, position, text } = selection;
+		return { sortIndex, pageLabel, position, text };
+	}
```

Comment:
I think it's too fragile to use private methods and variables this way, and if we want to get something from the reader or its views, we have to implement proper public methods in all views and the main reader API in reader.js, to get specific selection data that we want.

Also `_getAnnotationFromSelectionRanges` and `_getAnnotationFromTextSelection` just create an annotation from selection data, so I'm not sure we can call this selection.

What exactly plugin developers want from the selection? Do they want to position a popup? Or do they just need text? Annotation?

Reply:
The general need can be concluded as:

1. get selected text.
2. get a selection position, mainly for jumping back to the selection position
3. creating custom annotation from the selected text

Reply:
Aren't plugin developers also interested in listening evens i.e. when text is selected or selection changes? Are they just listening for pointerup events instead?

Reply:
It would be good if we add that. currently they use pointerup

Reply:
I suppose pointerup has to be added to the iframes of both views, otherwise they won't receive the event. And the second view (split view) and be created/destroyed at any time, which means plugins have to constantly check that and re-attach event listeners. Is this how plugins currently work? If so, we definitely need to make this simpler.

Selection can also be modified with Shift + arrow keys, even though this is currently not functioning in PDF view.

Reply:
yes. a mutation observer on the container element of the views to dynamically watch newly added views.

Reply:
Yes, I was also thinking about that. Allowing to hook into selection popup where plugin receives selection data and a callback that appends a custom HTML section into selection popup. This way multiple plugins could work together and can be implemented easily. @dstillman does it sound good?

I suppose getting the current selection is basically the most important thing that plugins want from the reader. What else do plugins need from the reader?

Reply:
Some plugins may also want to:
1. add a left-sidebar tab panel, alongside the annotations;
2. add buttons to the annotations in the annotation tab panel;
3. add buttons to the top-side tools bar.